syntax = "proto3";

service Hub {
    rpc Connect(stream Message) returns (stream Message);

    rpc PushTask(PushTaskRequest) returns (PushTaskResponse);
    rpc GetAgentsList(GetAgentsListRequest) returns (GetAgentsListResponse);
}

/****************************************************

Main concepts:
- Pull-based agent communacation
- Master (distributed due to fault tolerance) -> Wrokers (unstable services) model
- gRPC transport
- Fault tolerance scheduler


Requirements:

- Working with multiple same type workers
- Dynamic worker append
- Task cancellation
- Task progress checking

*****************************************************/

// ---- ---- ---- ---- Agents - Hub communication API ---- ---- ---- ---- //

enum Rpc {
    RPC_UNDEFINED = 0;
    RPC_REGISTER_AGENT = 1; // Agent
    RPC_HEARTBEAT_AGENT = 2; // Agent
    RPC_START_TASK = 3; // Hub
    RPC_ABORT_TASK = 4; // Hub
    RPC_UPDATE_TASK = 5; // Agent
}

message RequestMessage {
    Rpc rpc = 1;
    string payload = 3;
}

message ResponseMessage {
    optional string payload = 2;
    optional string error = 3;
}

message Message {
    string reply_to = 1;
    oneof message {
        RequestMessage request = 2;
        ResponseMessage response = 3;
    }
}

// ---- ---- ---- ---- Hub internal communication API ---- ---- ---- ---- //

message PushTaskRequest {
    string agent_id = 1;
    string method = 2;
    string params = 3;
    string payload = 4;
}

message PushTaskResponse {
    optional string task_id = 1;
    optional string error = 2;
}

message GetAgentsListRequest { }

message Input {
    string name = 1;
    string type = 2;
}

message Method {
    string name = 1;
    repeated Input inputs = 2;
}

message Agent {
    string id = 1;
    string name = 2;
    repeated Method methods = 3;
}

message GetAgentsListResponse {
    repeated Agent agents = 1;
}
